packages/components/src/navigation/use-create-navigation-tree.js

```diff
@@ -18,6 +23,47 @@ export const useCreateNavigationTree = () => {
 		removeNode: removeMenu,
 	} = useNavigationTreeNodes();
 
+	/**
+	 * Stores direct nested menus of menus
+	 * This makes it easy to traverse menu tree
+	 *
+	 * Key is the menu prop of the menu
+	 * Value is an array of menu keys
+	 */
+	const [ directSubMenus, setDirectSubMenus ] = useState( {} );
+	const getDirectSubMenus = ( menu ) => directSubMenus[ menu ] || [];
+
+	const traverseMenu = ( startMenu, callback ) => {
```

Comment:
Nice! Just to clarify my mental model, this looks like a breadth first search for graph traversal (because of the queue and the visited array). Is that correct?

Reply:
Yep, it's breadth-first traversal.

Reply:
I don't think there was a needed to keep a list of visited nodes here. It's useful in graph BFS to prevent infinite loops when there are cycles, but in this case it looks like the menu is always going to be a tree (since its structure is defined by React's component tree).

Reply:
I don't think there was a needed to keep a list of visited nodes here. It's useful in graph BFS to prevent infinite loops when there are cycles, but in this case it looks like the menu is always going to be a tree (since its structure is defined by React's component tree).